services/src/ogc/wms/request.rs

```diff
@@ -0,0 +1,199 @@
+use serde::{Deserialize, Serialize};
+use crate::util::from_str;
+use geoengine_datatypes::primitives::{BoundingBox2D, Coordinate2D, TimeInterval};
+use serde::de::Error;
+
+// TODO: ignore case for field names
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+#[serde(tag = "request")]
+#[allow(clippy::pub_enum_variant_names, clippy::large_enum_variant)]
+pub enum WMSRequest {
+    GetCapabilities(GetCapabilities),
+    GetMap(GetMap),
+    GetFeatureInfo(GetFeatureInfo),
+    GetStyles(GetStyles),
+    GetLegendGraphic(GetLegendGraphic),
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetCapabilities {
+    pub version: Option<String>,
+    pub service: String,
+    pub format: Option<String>, // TODO: Option<GetCapabilitiesFormat>,
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub enum GetCapabilitiesFormat {
+    TextXml //, ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetMap {
+    pub version: String,
+    #[serde(deserialize_with = "from_str")]
+    pub width: u32,
+    #[serde(deserialize_with = "from_str")]
+    pub height: u32,
+    #[serde(deserialize_with = "parse_bbox")]
+    pub bbox: BoundingBox2D,
+    pub format: GetMapFormat,
+    pub layer: String,
+    pub crs: String,  // TODO: parse CRS
+    pub styles: String,
+    #[serde(default)]
+    #[serde(deserialize_with = "parse_time")]
+    pub time: Option<TimeInterval>,
+    // TODO: parse Option<TimeInterval>,
+    pub transparent: Option<bool>,
+    pub bgcolor: Option<String>,
+    pub sld: Option<String>,
+    pub sld_body: Option<String>,
+    pub elevation: Option<String>,
+    pub exceptions: Option<String>, // TODO: parse Option<GetMapExceptionFormat>
+    // DIM_<name>
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub enum GetMapExceptionFormat {
+    TextXML //, ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub enum GetMapFormat {
+    #[serde(rename = "image/png")]
+    ImagePng //, ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetFeatureInfo {
+    pub version: String,
+    pub query_layers: String,
+    pub info_format: Option<String>, // TODO: parse Option<GetFeatureInfoFormat>,
+    // ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub enum GetFeatureInfoFormat {
+    TextXML //, ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetStyles {
+    pub version: String,
+    pub layer: String,
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetLegendGraphic {
+    pub version: String,
+    pub layer: String,
+    // ...
+}
+
+pub fn parse_bbox<'de, D>(deserializer: D) -> Result<BoundingBox2D, D::Error>
+    where D: serde::Deserializer<'de>
+{
+    // format is : "x1,y1,x2,y2"
```

Comment:
You can put it to the function docs

Reply:
b787cd17b9e41a9b2f4d52d58910f135bc0eb95e